File: MoviesPreview/app/src/main/java/com/jpp/moviespreview/core/mvp/StateFragment.java
package com.jpp.moviespreview.core.mvp;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;

/**
 * Created by jpperetti on 2/12/16.
 */

public class StateFragment<T extends BasePresenter<?>> extends Fragment {

    static final String TAG = StateFragment.class.getName();


    private BasePresenter<?> mPresenter;

    public static StateFragment newInstance() {
        return new StateFragment();
    }

    /**
     * Class constructor
     */
    public StateFragment() {

    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setRetainInstance(true);
    }


    public void setPresenter(@NonNull BasePresenter<?> presenter) {
        mPresenter = presenter;
    }


    @Nullable
    public BasePresenter<?> getPresenter() {
        return mPresenter;
    }

}

File: MoviesPreview/app_kotlin/src/main/java/com/jpp/moviespreview/ui/mvp/StateFragment.kt
package com.jpp.moviespreview.ui.mvp

import android.os.Bundle
import android.support.v4.app.Fragment

/**
 * Fragment used to save the Presenter state when the activity rotates.
 *
 * Created by jpp on 6/19/17.
 */
class StateFragment<V : PresentingView, T : BasePresenter<V>> : Fragment() {

    /**
     * Complains with the contract of Fragment to declare non-params constructor
     * with factory method.
     */
    companion object {
        var TAG = StateFragment::class.simpleName
    }

    var mPresenter: T? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        retainInstance = true
    }
}